internal/middleware/middleware.go

```diff
@@ -0,0 +1,128 @@
+package middleware
+
+import (
+	"fmt"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/gofiber/fiber/v2"
+	"github.com/gofiber/fiber/v2/middleware/cors"
+	"github.com/gofiber/fiber/v2/middleware/logger"
+	"github.com/vedhavyas/go-subkey/v2"
+	"github.com/vedhavyas/go-subkey/v2/ed25519"
+	"github.com/vedhavyas/go-subkey/v2/sr25519"
+)
+
+// Logger returns a logger middleware
+func Logger() fiber.Handler {
+	return logger.New()
+}
+
+// CORS returns a CORS middleware
+func CORS() fiber.Handler {
+	return cors.New()
+}
+
+// AuthMiddleware is a middleware that validates the authentication credentials
+func AuthMiddleware(challengeWindow int64) fiber.Handler {
+	return func(c *fiber.Ctx) error {
+		clientID := c.Get("X-Client-ID")
+		signature := c.Get("X-Signature")
+		challenge := c.Get("X-Challenge")
+
+		if clientID == "" || signature == "" || challenge == "" {
+			return c.Status(fiber.StatusUnauthorized).JSON(fiber.Map{
+				"error": "Missing authentication credentials",
+			})
+		}
+
+		// Verify the clientID and signature here
```

Comment:
Shouldn't we verify the signature first here before the validation of the challenge?

Reply:
IMO, it makes sense to validate first. For example, it may be unnecessary to verify the signature if it has already expired.

Reply:
When you receive an email do you check for the sender or the content first?  :)